cmd/clusterctl/internal/migrate/converter.go

```diff
@@ -0,0 +1,166 @@
+type gvkConversionMap map[schema.GroupVersionKind]schema.GroupVersionKind
+
+// ConversionResult represents the outcome of converting a single resource.
+type ConversionResult struct {
+	Object runtime.Object
+	// Converted indicates whether the object was actually converted
+	Converted bool
+	Error     error
+	Warnings  []string
+}
+
+// NewConverter creates a new resource converter using the clusterctl scheme.
+func NewConverter(targetGV schema.GroupVersion) (*Converter, error) {
+	return &Converter{
+		scheme:       scheme.Scheme,
+		targetGV:     targetGV,
+		targetGVKMap: make(gvkConversionMap),
+	}, nil
+}
+
+// ConvertResource converts a single resource to the target version.
+// Returns the converted object, or the original if no conversion is needed.
+func (c *Converter) ConvertResource(info ResourceInfo, obj runtime.Object) ConversionResult {
+	gvk := info.GroupVersionKind
+
+	if gvk.Group == clusterv1.GroupVersion.Group && gvk.Version == c.targetGV.Version {
+		return ConversionResult{
+			Object:    obj,
+			Converted: false,
+			Warnings:  []string{fmt.Sprintf("Resource %s/%s is already at version %s", gvk.Kind, info.Name, c.targetGV.Version)},
+		}
+	}
+
+	if gvk.Group != clusterv1.GroupVersion.Group {
+		return ConversionResult{
+			Object:    obj,
+			Converted: false,
+			Warnings:  []string{fmt.Sprintf("Skipping non-%s resource: %s", clusterv1.GroupVersion.Group, gvk.String())},
+		}
+	}
+
+	targetGVK, err := c.getTargetGVK(gvk)
+	if err != nil {
+		return ConversionResult{
+			Object:    obj,
+			Converted: false,
+			Error:     errors.Wrapf(err, "failed to determine target GVK for %s", gvk.String()),
+		}
+	}
+
+	// Check if the object is already typed
+	// If it's typed and implements conversion.Convertible, use the custom ConvertTo method
+	if convertible, ok := obj.(conversion.Convertible); ok {
+		// Create a new instance of the target type
+		targetObj, err := c.scheme.New(targetGVK)
+		if err != nil {
+			return ConversionResult{
+				Object:    obj,
+				Converted: false,
+				Error:     errors.Wrapf(err, "failed to create target object for %s", targetGVK.String()),
+			}
+		}
+
+		// Check if the target object is a Hub
+		if hub, ok := targetObj.(conversion.Hub); ok {
+			if err := convertible.ConvertTo(hub); err != nil {
+				return ConversionResult{
+					Object:    obj,
+					Converted: false,
+					Error:     errors.Wrapf(err, "failed to convert %s from %s to %s", gvk.Kind, gvk.Version, c.targetGV.Version),
+				}
+			}
+
+			// Ensure the GVK is set on the converted object
+			hubObj := hub.(runtime.Object)
+			hubObj.GetObjectKind().SetGroupVersionKind(targetGVK)
+
+			return ConversionResult{
+				Object:    hubObj,
+				Converted: true,
+				Error:     nil,
+				Warnings:  nil,
+			}
+		}
+	}
+
+	// Use scheme-based conversion for all remaining cases
+	convertedObj, err := c.scheme.ConvertToVersion(obj, targetGVK.GroupVersion())
```

Comment:
I would say we should fail hard if we are dealing with a core object that does not support convertible